--- a/config.h
+++ b/config.h
@@ -43,11 +43,16 @@
 			 */
 
 #define MAILCMD _PATH_SENDMAIL					/*-*/
-#define MAILARGS "%s -FCronDaemon -odi -oem -or0s %s"		/*-*/
-			/* -Fx	 = set full-name of sender
+/* #define MAILARGS "%s -i -FCronDaemon -odi -oem  %s"		/*-*/
+#define MAILARGS "%s -i -FCronDaemon -oem  %s"		/*-*/
+			/* -i    = don't terminate on "." by itself
+                         * -Fx	 = set full-name of sender
 			 * -odi	 = Option Deliverymode Interactive
 			 * -oem	 = Option Errors Mailedtosender
+ 			 * -t    = read recipient from header of message
 			 * -or0s = Option Readtimeout -- don't time out
+			 * XXX: sendmail doesn't allow -or0s when invoked
+			 * by joe user.  --okir
 			 */
 
 /* #define MAILCMD "/bin/mail"			/*-*/
