--- a/cron.c
+++ b/cron.c
@@ -71,6 +71,7 @@ main(argc, argv)
 	char	*argv[];
 {
 	cron_db	database;
+	char *cs;
 
 	ProgramName = argv[0];
 
@@ -96,6 +97,19 @@ main(argc, argv)
 	setenv("PATH", _PATH_DEFPATH, 1);
 #endif
 
+       /* Get the default locale character set for the mail
+        * "Content-Type: ...; charset=" header
+        */
+       setlocale(LC_ALL,""); /* set locale to system defaults or to
+                                that specified by any  LC_* env vars */
+       /* Except that "US-ASCII" is preferred to "ANSI_x3.4-1968" in MIME,
+        * even though "ANSI_x3.4-1968" is the official charset name. */
+       if ( ( cs = nl_langinfo( CODESET ) ) != 0L && 
+               strcmp(cs, "ANSI_x3.4-1968") != 0 )
+           strncpy( cron_default_mail_charset, cs, MAX_ENVSTR );
+       else
+           strcpy( cron_default_mail_charset, "US-ASCII" );
+
 	/* if there are no debug flags turned on, fork as a daemon should.
 	 */
 # if DEBUGGING
