--- a/cron.c
+++ b/cron.c
@@ -421,7 +421,7 @@ parse_args(argc, argv)
 
 	stay_foreground = 0;
 
-	while (EOF != (argch = getopt(argc, argv, "lfx:"))) {
+	while (EOF != (argch = getopt(argc, argv, "lfx:L:"))) {
 		switch (argch) {
 		default:
 			usage();
@@ -435,6 +435,9 @@ parse_args(argc, argv)
                 case 'l':
 			lsbsysinit_mode = 1;
 			break;
+		case 'L':
+			log_level = atoi(optarg);
+			break;
 		}
 	}
 }
--- a/cron.h
+++ b/cron.h
@@ -272,6 +272,7 @@ time_min clockTime;
 int	stay_foreground;
 
 int	lsbsysinit_mode;
+int	log_level = 1;
 
 # if DEBUGGING
 int	DebugFlags;
@@ -286,6 +287,7 @@ extern	char	*copyright[],
 		*DowNames[],
 		*ProgramName;
 extern	int	lsbsysinit_mode;
+extern	int	log_level;
 extern	int	LineNumber;
 extern	time_t	StartTime;
 extern  time_min timeRunning;
